test/Microsoft.AspNetCore.SignalR.Tests/HubEndpointTests.cs

```diff
@@ -1333,6 +1333,78 @@ public static IEnumerable<object[]> StreamingMethodAndHubProtocols
             }
         }
 
+        [Fact]
+        public async Task DoesNotWritePingMessagesIfSufficientOtherMessagesAreSent()
+        {
+            var serviceProvider = CreateServiceProvider(services =>
+                services.Configure<HubOptions>(options =>
+                    options.KeepAliveInterval = TimeSpan.FromMilliseconds(100)));
+            var endPoint = serviceProvider.GetService<HubEndPoint<MethodHub>>();
+
+            using (var client = new TestClient(false, new JsonHubProtocol()))
+            {
+                var endPointLifetime = endPoint.OnConnectedAsync(client.Connection).OrTimeout();
+
+                await client.Connected.OrTimeout();
+
+                // Echo a bunch of stuff, waiting 10ms between each, until 500ms have elapsed
+                DateTime start = DateTime.UtcNow;
+                while ((DateTime.UtcNow - start).TotalMilliseconds <= 500.0)
+                {
+                    await client.SendInvocationAsync("Echo", "foo").OrTimeout();
+                    await Task.Delay(10);
+                }
+
+                // Shut down
+                client.Dispose();
+
+                await endPointLifetime.OrTimeout();
+
+                // We shouldn't have any ping messages
+                HubMessage message;
+                var counter = 0;
+                while ((message = await client.ReadAsync()) != null)
+                {
+                    counter += 1;
+                    Assert.IsNotType<PingMessage>(message);
+                }
+                Assert.InRange(counter, 1, 50);
+            }
+        }
+
+        [Fact]
+        public async Task WritesPingMessageIfNothingWrittenWhenKeepAliveIntervalElapses()
+        {
+            var serviceProvider = CreateServiceProvider(services =>
+                services.Configure<HubOptions>(options =>
+                    options.KeepAliveInterval = TimeSpan.FromMilliseconds(100)));
+            var endPoint = serviceProvider.GetService<HubEndPoint<MethodHub>>();
+
+            using (var client = new TestClient(false, new JsonHubProtocol()))
+            {
+                var endPointLifetime = endPoint.OnConnectedAsync(client.Connection).OrTimeout();
+                await client.Connected.OrTimeout();
+
+                // Wait 500 ms
+                await Task.Delay(500);
+
+                // Shut down
+                client.Dispose();
+
+                await endPointLifetime.OrTimeout();
+
+                // We should have all pings
+                HubMessage message;
+                var counter = 0;
+                while ((message = await client.ReadAsync()) != null)
+                {
+                    counter += 1;
+                    Assert.Same(PingMessage.Instance, message);
+                }
+                Assert.InRange(counter, 1, Int32.MaxValue);
```

Comment:
1 - 5

Reply:
I'm going to do 1-10, because I want to make sure this doesn't flake out occasionally due to random delays here and there. Remember that `Task.Delay` only guarantees that you won't be woken up before the delay has elapsed; you might be sleeping for longer than the delay.